// react-magic-cursor/dist/index.es.js
"use client";
import F, { createContext as _, useContext as W, useCallback as m, Children as q, cloneElement as B, useState as C, useLayoutEffect as z, useRef as v, useEffect as R, useMemo as b } from "react";
const S = "cursor-hover", y = 30, D = 30, h = "#000000", f = "outline", M = 9999, V = 1, I = 0, j = 0, U = 8, N = _(null), $ = _(null), G = () => W(N), K = () => W($), k = ({
  children: r,
  type: c = f,
  color: e = h,
  offset: t = I,
  className: n = "",
  ...u
}) => {
  const { outlineElement: s, underlineElement: l, leaveElement: o } = K(), d = m(
    (a) => {
      if (!a.currentTarget)
        return;
      const p = a.currentTarget.getBoundingClientRect(), x = getComputedStyle(a.currentTarget);
      c === f ? s(p, x, e, t) : l(p, e);
    },
    [s, l, e, t, c]
  ), L = m(() => {
    o();
  }, [o]);
  return q.map(
    r,
    (a) => B(a, {
      onMouseEnter: d,
      onMouseLeave: L,
      className: `${n} ${a.props.className || ""}`.trim(),
      ...u
    })
  );
}, J = () => {
  const [r, c] = C({
    scrollX: 0,
    scrollY: 0
  });
  return z(() => {
    if (!window)
      return;
    const e = () => {
      c({ scrollX: window.scrollX, scrollY: window.scrollY });
    };
    return window.addEventListener("scroll", e), () => window.removeEventListener("scroll", e);
  }, []), r;
}, Q = () => {
  const [r, c] = C({ x: 0, y: 0 }), { scrollX: e, scrollY: t } = J();
  return z(() => {
    if (!window)
      return;
    const n = ({ clientX: u, clientY: s }) => {
      c({ x: u + e, y: s + t });
    };
    return window.addEventListener("pointermove", n), () => window.removeEventListener("pointermove", n);
  }, [e, t]), r;
}, rr = () => {
  const { cursorRef: r, cursorFrozenPosition: c, cursorSize: e, cursorStyles: t } = G(), n = Q(), u = v(null), s = v(null), l = v(null), o = n;
  return R(() => {
    r.current && (u.current = c || n, s.current = e, Object.assign(r.current.style, {
      ...t,
      width: `${s.current.width}px`,
      height: `${s.current.height}px`,
      borderWidth: `${t.borderWidth}px`,
      borderRadius: `${t.borderRadius}px`
    }));
  }, [r, n, c, e, t]), R(() => {
    const d = () => {
      r.current && (o.x += (u.current.x - o.x) / U, o.y += (u.current.y - o.y) / U, Object.assign(r.current.style, {
        top: `${o.y}px`,
        left: `${o.x}px`
      }), l.current = requestAnimationFrame(d));
    };
    return l.current = requestAnimationFrame(d), () => {
      l.current && cancelAnimationFrame(l.current);
    };
  }, []), /* @__PURE__ */ F.createElement("div", { className: "cursor", ref: r, "data-testid": "cursor" });
}, er = ({
  thickness: r = V,
  children: c
}) => {
  const e = v(null), [t, n] = C(null), [u, s] = C({
    width: y,
    height: D
  }), [l, o] = C({
    borderWidth: r,
    borderRadius: M,
    borderColor: h
  }), d = m(
    (i, E, P = h, w = I) => {
      if (!e.current)
        return;
      const { borderRadius: g, borderWidth: O } = E, Y = +g.replace("px", ""), X = +O || 0, A = w + r / 2, T = X + w * 2;
      e.current.classList.add(S), n({
        x: i.x + window.scrollX - A,
        y: i.y + window.scrollY - A
      }), s({
        width: i.width + T,
        height: i.height + T
      }), o((H) => ({
        ...H,
        borderRadius: Y + r + w * 2 / Math.PI,
        borderColor: P
      }));
    },
    [r]
  ), L = m(
    (i, E = h, P = j) => {
      if (!e.current)
        return;
      const w = r / 2;
      e.current.classList.add(S), n({
        x: i.x + window.scrollX,
        y: i.y + i.height + window.scrollY + P
      }), s({
        width: i.width,
        height: 0
      }), o((g) => ({
        ...g,
        borderWidth: w,
        borderColor: E
      }));
    },
    [r]
  ), a = m(() => {
    var i;
    n(null), s({
      height: D,
      width: y
    }), o((E) => ({
      ...E,
      borderWidth: r,
      borderRadius: M,
      borderColor: h
    })), (i = e.current) == null || i.classList.remove(S);
  }, [r]), p = b(
    () => ({
      cursorRef: e,
      cursorFrozenPosition: t,
      cursorSize: u,
      cursorStyles: l
    }),
    [e, t, u, l]
  ), x = b(
    () => ({
      outlineElement: d,
      underlineElement: L,
      leaveElement: a
    }),
    [d, L, a]
  );
  return /* @__PURE__ */ F.createElement(N.Provider, { value: p }, /* @__PURE__ */ F.createElement($.Provider, { value: x }, c));
};
export {
  rr as MagicCursor,
  er as MagicCursorProvider,
  k as MagicElement
};
